File: manual/tools/formats.py
"""Authored format catalog loading, validation, and scaffold defaults."""

from pathlib import Path
import re

import yaml

import schema_validation


MANUAL = Path(__file__).resolve().parents[1]
LEGACY_FORMAT_ROUTES = {
    "teamtypes": "/mapping/team-types/",
    "taskforces": "/mapping/task-forces/",
    "scripts": "/mapping/scripts/",
    "ai_triggers": "/mapping/ai-triggers/",
}
FORMAT_KINDS = {"syntax", "file", "registry", "record", "binary"}
BINARY_ROLES = {"archive", "image", "model", "audio", "video", "persistence"}
_FRONTMATTER = re.compile(r"^---\r?\n([\s\S]*?)\r?\n---")


def _frontmatter_text(source, context):
    match = _FRONTMATTER.match(source)
    if not match:
        raise ValueError(f"{context}: missing YAML frontmatter")
    data = yaml.safe_load(match.group(1)) or {}
    if not isinstance(data, dict):
        raise ValueError(f"{context}: frontmatter must be a mapping")
    return data


def format_route(record, slug):
    """Return the canonical public route for an authored format."""
    route = record.get("route")
    if not route:
        return f"/formats/{slug}/"
    expected = LEGACY_FORMAT_ROUTES.get(record.get("format_id"))
    if expected is None:
        raise ValueError(
            "route override is reserved for the four migrated AI format records")
    if route != expected:
        raise ValueError(
            f"route for {record.get('format_id')!r} must remain {expected!r}")
    return route


def parse_catalog(documents):
    """Parse ``(path, Markdown source)`` pairs into a catalog keyed by format ID.

    This pure boundary is also used for Git-revision snapshots. Paths may be
    absolute or repository-relative; only their filename stems affect routes.
    """
    records = {}
    routes = {}
    for path, source in documents:
        candidate = Path(path)
        data = _frontmatter_text(source, str(candidate))
        format_id = data.get("format_id")
        if not isinstance(format_id, str) or not format_id:
            continue
        if format_id in records:
            raise ValueError(f"{candidate}: duplicate format ID {format_id!r}")
        normalized = dict(data)
        normalized.setdefault("related", [])
        normalized.setdefault("key_scopes", [])
        if normalized.get("kind") == "binary":
            normalized.setdefault("companion_formats", [])
        normalized["route"] = format_route(normalized, candidate.stem)
        normalized["_slug"] = candidate.stem
        owner = routes.get(normalized["route"])
        if owner:
            raise ValueError(
                f"{candidate}: route {normalized['route']} is already owned by {owner}")
        routes[normalized["route"]] = format_id
        records[format_id] = normalized
    return records


def load_formats(manual=MANUAL):
    """Load the active authored Format catalog keyed by stable ``format_id``."""
    folder = Path(manual) / "content" / "formats"
    documents = (
        (path, path.read_text(encoding="utf-8"))
        for path in sorted(folder.glob("*.md"))
    )
    return parse_catalog(documents)


def _selector_matches(left, right):
    return isinstance(left, dict) and left == right


def _query_matches(scope, query):
    return (
        (not query.get("file") or scope.get("file") == query["file"])
        and (not query.get("section")
             or _selector_matches(scope.get("section"), query["section"]))
        and (not query.get("applies_to")
             or query["applies_to"] in scope.get("applies_to", []))
    )


def _contains_todo(value):
    if isinstance(value, str):
        return value.startswith("TODO:")
    if isinstance(value, dict):
        return any(_contains_todo(item) for item in value.values())
    if isinstance(value, list):
        return any(_contains_todo(item) for item in value)
    return False


def validate_formats(errors, keys=None, manual=MANUAL):
    """Validate authored formats and return their normalized lifecycle catalog."""
    manual = Path(manual)
    folder = manual / "content" / "formats"
    records = {}
    routes = {}
    source_documents = []

    for path in sorted(folder.glob("*.md")):
        label = str(path.relative_to(manual))
        source = path.read_text(encoding="utf-8")
        try:
            data = _frontmatter_text(source, label)
        except (OSError, ValueError, yaml.YAMLError) as error:
            errors.append(str(error))
            continue
        errors.extend(schema_validation.errors_for(
            data, "authored-format.schema.json", label))
        if _contains_todo(data):
            errors.append(
                f"{label}: replace all TODO placeholders before publishing this format")
        format_id = data.get("format_id")
        if not isinstance(format_id, str) or not format_id:
            continue
        if format_id in records:
            errors.append(f"{label}: duplicate format ID {format_id!r}")
            continue

        try:
            route = format_route(data, path.stem)
        except ValueError as error:
            errors.append(f"{label}: {error}")
            route = f"/formats/{path.stem}/"
        route_tail = next((part for part in reversed(route.split("/")) if part), "")
        if route_tail != path.stem:
            errors.append(
                f"{label}: filename must be {route_tail}.md for canonical route {route}")
        if route in routes:
            errors.append(
                f"{label}: route {route} is already owned by {routes[route]}")
        else:
            routes[route] = format_id

        normalized = dict(data)
        normalized.setdefault("related", [])
        normalized.setdefault("key_scopes", [])
        if normalized.get("kind") == "binary":
            normalized.setdefault("companion_formats", [])
        normalized["route"] = route
        normalized["_slug"] = path.stem
        records[format_id] = normalized
        source_documents.append((path, label, normalized))

    for path, label, record in source_documents:
        for source_file in record.get("source_files", []):
            if not isinstance(source_file, str):
                continue
            candidate = Path(source_file)
            if candidate.is_absolute() or ".." in candidate.parts:
                errors.append(f"{label}: invalid source path {source_file!r}")
            elif not (manual.parent / candidate).is_file():
                errors.append(f"{label}: source path does not exist: {source_file}")

        fields = record.get("fields", [])
        positions = [field.get("position") for field in fields if isinstance(field, dict)]
        if positions and positions != list(range(1, len(positions) + 1)):
            errors.append(
                f"{label}: record fields must use ordered consecutive positions starting at 1")

        registrations = record.get("registrations", [])
        registration_sections = [
            item.get("section") for item in registrations if isinstance(item, dict)
        ]
        if len(registration_sections) != len(set(registration_sections)):
            errors.append(f"{label}: duplicate registration section")

        for companion in record.get("companion_formats", []):
            if companion == record.get("format_id"):
                errors.append(f"{label}: a format cannot be its own companion")
            elif companion not in records:
                errors.append(f"{label}: unknown companion format {companion!r}")

        if keys is not None:
            scopes = [
                scope
                for entry in keys.values() if isinstance(entry, dict)
                for scope in entry.get("scopes", []) if isinstance(scope, dict)
            ]
            for position, query in enumerate(record.get("key_scopes", []), start=1):
                if isinstance(query, dict) and not any(
                        _query_matches(scope, query) for scope in scopes):
                    errors.append(
                        f"{label}: key_scopes entry {position} matches no generated scope")

    return records


def scaffold_frontmatter(format_id, kind):
    """Return minimal frontmatter for ``scaffold format``.

    TODO values intentionally keep validation/action-required output active until
    the programmer supplies source-backed facts.
    """
    if kind not in FORMAT_KINDS:
        raise ValueError(
            f"unknown format kind {kind!r}; expected one of {', '.join(sorted(FORMAT_KINDS))}")
    result = {
        "format_id": format_id,
        "title": "TODO: replace with the display title",
        "summary": "TODO: replace with one literal sentence",
        "kind": kind,
        "source_files": ["TODO: replace with a repository-relative source path"],
    }
    if kind == "syntax":
        result["applies_to"] = ["TODO: replace with the affected inputs"]
    elif kind == "file":
        result["filenames"] = ["TODO: replace with the accepted filename"]
    elif kind == "registry":
        result["files"] = ["TODO: replace with the containing file"]
        result["registrations"] = [{
            "section": "TODO: replace with the registration section",
            "id_from": "value",
            "entry_section": "TODO: replace with the definition section",
        }]
    elif kind == "record":
        result["files"] = ["TODO: replace with the containing file"]
        result["section"] = "TODO: replace with the containing section"
        result["syntax"] = "TODO: replace with the record syntax"
        result["fields"] = [{
            "position": 1,
            "label": "TODO: replace with the field label",
            "value": "TODO: replace with the accepted value",
            "required": True,
        }]
    else:
        result["extensions"] = [".TODO"]
        result["role"] = "TODO: choose archive, image, model, audio, video, or persistence"
    return result

File: manual/site/src/lib/formats.ts
/*
 * Authored format catalog. Metadata is loaded synchronously so routes,
 * lifecycle links, and cross-references can resolve formats without coupling
 * the generated-data layer to Astro's asynchronous content API.
 */

import { load } from 'js-yaml';
import authoredFormat from '../../../schema/authored-format.schema.json';
import { UI } from '../i18n/en.mjs';
import type { KeyRecord, KeyScope } from './data';
import { assertContract } from './schema-validator';
import type { SectionSelector } from './section-selector.mjs';

export type FormatKind = 'syntax' | 'file' | 'registry' | 'record' | 'binary';

export interface FormatReference {
	type: 'key' | 'action' | 'event' | 'mission' | 'format' | 'enum' | 'system' | 'command' | 'guide' | 'using' | 'internal';
	id: string;
	scope?: string;
}

export interface FormatKeyScopeQuery {
	file?: string;
	section?: SectionSelector;
	applies_to?: string;
}

export interface FormatRegistration {
	section: string;
	id_from: 'key' | 'value' | 'entry-name';
	entry_section?: string;
	value?: string;
}

export interface FormatField {
	position: number;
	label: string;
	value: string;
	required: boolean;
	note?: string;
}

interface FormatCommon {
	format_id: string;
	title: string;
	summary: string;
	kind: FormatKind;
	route?: string;
	source_files: string[];
	related: FormatReference[];
	key_scopes: FormatKeyScopeQuery[];
}

export interface SyntaxFormat extends FormatCommon {
	kind: 'syntax';
	applies_to: string[];
}

export interface FileFormat extends FormatCommon {
	kind: 'file';
	filenames: string[];
}

export interface RegistryFormat extends FormatCommon {
	kind: 'registry';
	files: string[];
	registrations: FormatRegistration[];
}

export interface RecordFormat extends FormatCommon {
	kind: 'record';
	files: string[];
	section: string;
	syntax: string;
	registration?: FormatRegistration;
	enable_section?: string;
	fields: FormatField[];
}

export interface BinaryFormat extends FormatCommon {
	kind: 'binary';
	extensions: string[];
	role: 'archive' | 'image' | 'model' | 'audio' | 'video' | 'persistence';
	companion_formats: string[];
}

export type FormatData = SyntaxFormat | FileFormat | RegistryFormat | RecordFormat | BinaryFormat;

export type FormatRecord = FormatData & {
	slug: string;
	route: string;
};

export interface FormatKeyMatch {
	record: KeyRecord;
	scope: KeyScope;
}

const RAW_FORMATS = import.meta.glob('../../../content/formats/*.md', {
	query: '?raw',
	import: 'default',
	eager: true,
}) as Record<string, string>;

let formatCache: Map<string, FormatRecord> | null = null;

function parseFrontmatter(path: string, source: string): Record<string, unknown> {
	const match = source.match(/^---\r?\n([\s\S]*?)\r?\n---/);
	if (!match) throw new Error(`${path}: missing format frontmatter`);
	const value = load(match[1]);
	assertContract(authoredFormat, value, path);
	return value as Record<string, unknown>;
}

function routeTail(route: string): string {
	const tail = route.split('/').filter(Boolean).at(-1);
	if (!tail) throw new Error(`Format route ${JSON.stringify(route)} has no path segment`);
	return tail;
}

function normalize(path: string, source: string): FormatRecord {
	const data = parseFrontmatter(path, source) as unknown as FormatData;
	const slug = path.split('/').at(-1)?.replace(/\.md$/, '');
	if (!slug) throw new Error(`${path}: cannot derive format slug`);
	const route = data.route ?? `/formats/${slug}/`;
	if (routeTail(route) !== slug) {
		throw new Error(`${path}: filename must match the final segment of ${route}`);
	}
	return {
		...data,
		related: data.related ?? [],
		key_scopes: data.key_scopes ?? [],
		...(data.kind === 'binary' ? { companion_formats: data.companion_formats ?? [] } : {}),
		slug,
		route,
	} as FormatRecord;
}

export function allFormats(): Map<string, FormatRecord> {
	if (formatCache) return formatCache;
	formatCache = new Map();
	const routes = new Map<string, string>();
	for (const [path, source] of Object.entries(RAW_FORMATS)) {
		const record = normalize(path, source);
		if (formatCache.has(record.format_id)) {
			throw new Error(`Duplicate format ID: ${record.format_id}`);
		}
		const routeOwner = routes.get(record.route);
		if (routeOwner) {
			throw new Error(`Format route ${record.route} is shared by ${routeOwner} and ${record.format_id}`);
		}
		routes.set(record.route, record.format_id);
		formatCache.set(record.format_id, record);
	}

	for (const record of formatCache.values()) {
		if (record.kind !== 'binary') continue;
		for (const companion of record.companion_formats) {
			if (!formatCache.has(companion)) {
				throw new Error(`Format ${record.format_id}: unknown companion format ${companion}`);
			}
			if (companion === record.format_id) {
				throw new Error(`Format ${record.format_id}: a format cannot be its own companion`);
			}
		}
	}
	return formatCache;
}

export function formatBySlug(slug: string): FormatRecord | undefined {
	return [...allFormats().values()].find((record) => record.slug === slug);
}

export function formatHref(formatId: string): string | undefined {
	return allFormats().get(formatId)?.route;
}

export function formatLabel(formatId: string): string | undefined {
	return allFormats().get(formatId)?.title;
}

function sameSelector(left: SectionSelector, right: SectionSelector): boolean {
	if (left.kind !== right.kind) return false;
	if (left.kind === 'literal' && right.kind === 'literal') return left.name === right.name;
	if (left.kind === 'identifier' && right.kind === 'identifier') return left.source === right.source;
	// An Image= read that keeps the object's own name as its fallback is a
	// different lookup from one that has no fallback, so a query for either
	// must never match the other.
	return left.kind === 'image' && right.kind === 'image'
		&& (left.fallback ?? null) === (right.fallback ?? null);
}

function matches(scope: KeyScope, query: FormatKeyScopeQuery): boolean {
	return (!query.file || scope.file === query.file)
		&& (!query.section || sameSelector(scope.section, query.section))
		&& (!query.applies_to || scope.applies_to.includes(query.applies_to));
}

/** Resolve accepted settings from generated scopes without copying key lists into Markdown. */
export function formatKeyMatches(format: FormatData, keys: ReadonlyMap<string, KeyRecord>): FormatKeyMatch[] {
	const matchesByIdentity = new Map<string, FormatKeyMatch>();
	for (const record of keys.values()) {
		for (const scope of record.scopes) {
			if (!format.key_scopes.some((query) => matches(scope, query))) continue;
			matchesByIdentity.set(`${record.key}\u0000${scope.route_id}`, { record, scope });
		}
	}
	return [...matchesByIdentity.values()].sort((left, right) =>
		left.record.key.localeCompare(right.record.key)
		|| left.scope.route_id.localeCompare(right.scope.route_id));
}

export function formatKindLabel(kind: FormatKind): string {
	const entry = UI.formatKinds.find((candidate) => candidate.id === kind);
	if (!entry) throw new Error(`Unknown format kind: ${kind}`);
	return entry.title;
}
